### ISSUE TITLE
Bug: [no-invalid-void-type] false positive (?) when used as `Promise<T | void>`

### ISSUE THREAD

**Issue by xsjcTony** [typescript-eslint/typescript-eslint#7032]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=MYewdgzgLgBATgUygVzmAylOBLMBzGAXhgEMIBPMYGACgEoAuGABThAFtsIEAeaHfAD4iw1hy4IAdIgggANgDcENAORQE0FXQCwAKFCRYiFGgBqIbABMipClVqMWbTtx4KLl4YVHOJ0jfJK9Hp6APShMO5WMFww4HLkkSRy0WSk8EioYDBQ5AAOCHFwMHgIYAg41LkFpHB4yOxlsCDFUAAWhdWFIABm6QAG7Vz9MHkkcCSN6nCSegbQGSYYWLh4APJw5tHEZJTUND1yJHhMAEYg8ggkYI5iLrz8qzAAPpEeXsIA3nowi1kwh2OP1+MAA-KQAO4kbBGTJoTACPDBXQgmBMEhQmF-MweZEAXxCujCEQAKvlCioopYVDEIDAwCBYGQINg8GASKc5J0QDlyTAVI98C8YMgwJYED1cAhqZJaAAmADMcrlOn04AWxiyCNWGwAqmKJVLrDs7PtAScYOdLtdbr5XIKCK9ReLJeVPCIYN8Udjsubgb9wRjobCltr8MjUejMSGsltLPi9EA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6MAG1w2HImjQO0SABp5CrAuyQB8DgHcAkkwDiiJkspkAKiUQBBaAHNYAWyv5UGfHERNXT0DYydkOwALSmQABT5oPi98JWl-WEQtbABfLIBdeVzsoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
const returnString = async (): Promise<string> => Promise.resolve('test')
const returnVoid = async (): Promise<void> => Promise.resolve()

// void is only valid as a return type or generic type argument or the type of a `this` parameter.
const returnStringOrVoid = async (flag: boolean): Promise<string | void> => {
  return flag
    ? await returnString()
    : await returnVoid()
}


// Type 'void' is not assignable to type 'string | undefined'. (2322)
const returnStringOrUndefined = async (flag: boolean): Promise<string | undefined> => {
  return flag
    ? await returnString()
    : await returnVoid()
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-invalid-void-type": [
      "error",
      {
        "allowInGenericTypeArguments": true,
        "allowAsThisParameter": true
      }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No error is thrown, since it respects the actual return type

### Actual Result

Raising error  `void is only valid as a return type or generic type argument or the type of a `this` parameter.`

**Comment by Josh-Cena** [comment]
My take is that anything unioned with `void` is a bad type to write, because `void` in the return position is basically `unknown`—for example, you can assign `() => number` to `() => void`, which means if you have `() => string | void`, you have no idea what that `void` actually represents at runtime—it doesn't have to be `undefined`. In this case you would be equally better off with a single `Promise<unknown>`.

**Comment by xsjcTony** [comment]
Yeah I'll leave it as `Promise<T | undefined>` anyway, although it's giving error now, but in TS5.1 it's allowing function that returns `undefined` not to have explicit `return` statement, so I'm looking forward to that.
